src/torrent.cpp

```diff
@@ -4392,9 +4392,9 @@ namespace {
 		if (m_ses.alerts().should_post<piece_finished_alert>())
 			m_ses.alerts().emplace_alert<piece_finished_alert>(get_handle(), index);
 
-		if (has_picker() && picker().is_piece_ready(index))
+		if (alerts().should_post<piece_ready_alert>())
 		{
-			if (alerts().should_post<piece_ready_alert>())
+			if (has_picker() && picker().is_piece_ready(index))
```

Comment:
I'm not sure under what circumstances we won't have a picker here. You think it would make sense to call need_picker() to ensure the alert is never missed?

Likewise on the `on_write_completed` callbacks would bail out before sending the alert if there's no picker. `on_piece_verified` calls `need_picker()`

Reply:
when the torrent becomes a seed, the picker is deleted (as there is no need to keep the additional state associated with picking pieces anymore).

If a picker is expected to always exist through some code path, it probably makes sense to add an assert to ensure that assumption holds